### PR TITLE
Make Sessions Recording faster!!

### LINKED ISSUE #6269: Slow session recording playback loading
## In what situation are you experiencing subpar performance?

Session recording playback loading times are slow because we're querying for a large amount of session recording events for long sessions. This delay starts getting noticeable when recordings > 1hr on my 32GB Macbook Pro, but I can imagine it gets worse for slower computers.

Should be fixed as part of #6077 epic

## How to reproduce

1. Click "Watch session" for a long session recording
2. Wait a long time until you can start watching the recording

## Environment

- [X] PostHog Cloud
- [X] self-hosted PostHog, version/commit: _please provide_

### PR DESCRIPTION
**Pull request by alexkim205**
## Changes

Fixes #6269

Today our longest session recordings take a long time to load (3hr = 4s on my M1 mac, sometimes hangs for minutes on older computers).

Super lofi side-by-side comparison (Before | After)

https://user-images.githubusercontent.com/13460330/136300361-a5fe0b7d-b247-4600-8b95-a31540917ebc.mov

This PR changes the way that we interact with today's single session recording endpoints. Instead of making a bulk request that captures and returns the entire recording (a 3hr recording roughly translates to X MB sent over the wire), we fetch the recording in chunks that are defined by `limit` and `offset` search query parameters set in the request. This allows us to reduce the time to first paint so that the user can start interacting with the player quickly. 

Today, it looks like there're two different ways to fetch the snapshots of a session recording in today's code. `/api/events/session_recording/?session_recording_id={id}` or `/api/projects/@current/session_recordings/{id}` (newer API as per @rcmarron's most recent work). I added this limit/offset functionality to both endpoints to maintain backwards compatibility. It looks like the frontend still uses the first endpoint in playback view— @rcmarron should all session recording pages eventually be using the second newer api?

I would love some feedback on the following:

- BACKEND: changes to API adding LIMIT and OFFSET parameters to SQL query
- BACKEND: refactor of the two versions of `SessionRecording` classes
- BACKEND: construction of `next` url
- BACKEND: proper testing of everything above
- FRONTEND: calling subsequent requests on frontend
- FRONTEND: tests
- general code quality

~~This PR requires changes from https://github.com/PostHog/posthog-react-rrweb-player/pull/13. Please wait until this one is merged before pulling this branch and testing locally, if you want to avoid weird `yarn link` symlink troubleshooting.~~ 

Just published a `v1.1.3` of rrweb player that's compatible with this PR. When you pull this branch locally, all you need to do is a fresh `yarn install` to grab the latest update from the player repo.

## How did you test this code?

- [x] Test new sequential api querying logic in backend
- [x] Add util test for `format_offset_absolute_url()`
- [x] Logic tests for loader logic

### CODE REVIEW

File: frontend/src/scenes/sessions/sessionsPlayLogic.ts [hunk 1]
```diff
@@ -138,12 +153,25 @@ export const sessionsPlayLogic = kea<sessionsPlayLogicType<SessionPlayerData, Se
...
         sessionPlayerData: {
-            loadRecording: async (sessionRecordingId: string): Promise<SessionPlayerData> => {
+            loadRecording: async ({ sessionRecordingId, url }: RecordingRequest): Promise<SessionPlayerData> => {
                 const startTime = performance.now()
-                const params = toParams({ session_recording_id: sessionRecordingId, save_view: true })
-                const response = await api.get(`api/event/session_recording?${params}`)
+
+                let response
+                if (url) {
+                    // See https://github.com/PostHog/posthog/blob/b015df3eeb8444d0a20f36e0479cca39cddf69d7/posthog/queries/sessions/session_recording.py#L76-L76
```
Comment (macobo) [inline frontend/src/scenes/sessions/sessionsPlayLogic.ts:161]: The link seems kind of useless. What's the goal here?

**Review by macobo — DISMISSED** [review]
PR description is lacking. I suggest adding:
1. What approach is taken
2. What should reviewers pay special attention to

Also given it's a draft but requesting for feedback it should be clear what's still TBD vs what's done?

Given this changes the loading model a lot and this is not at all obvious from the code:
1. What happens if network is slow and chunks arrive too slow?
2. What happens if requests fail due to network?

**Comment by alexkim205** [comment]
> PR description is lacking. I suggest adding:
> 
> 1. What approach is taken
> 2. What should reviewers pay special attention to
> 
> Also given it's a draft but requesting for feedback it should be clear what's still TBD vs what's done?
> 
> Given this changes the loading model a lot and this is not at all obvious from the code:
> 
> 1. What happens if network is slow and chunks arrive too slow?
> 2. What happens if requests fail due to network?

The implementation isn't completely there yet and there're quite a few things to cleanup but thanks for the early review! I definitely should flesh out the PR more before tagging for feedback. I'll update the description with some more info.

1. As soon as the first chunk loads, the user will be able to start playing the first part of the recording (opposed to having to wait for the entire recording to get queried and returned). If the n+1th chunk arrives after the user gets to that point in the recording, I'll add a loading overlay indicator to show that the rest of the recording is buffering.
2. If anyone of the request fail, I'll expose an error toast which is akin to how we're erroring out in the rest of the app. The sequential order of the request make it so that we won't fetch the next chunk unless the current chunk was successfully received, so the user will see at most one error per recording. I think it'd also help to automatically retry fetching the request 3(?) times if parts of the recording have already been received, since it could just be a transient error that can be recovered.

**Review by rcmarron — DISMISSED** [review]
Great work, it's definitely much faster!

A couple of notes:
* I did a rough review of the backend, but I imagine thing might change as a result of the `decompress_chunked_snapshot_data` comment, so I didn't get too detailed past that. Also, let me know if you have questions on this comment, I was just digging into this as a part of the payload size bug.
* For the question about the 2 endpoints, the idea there was that we would start pointing the player to the new endpoint, but I think it still needs some work to get it fully functional. The idea was to make deleting the old sessions stuff is much easier when the time comes. It might make sense to address that fully when we get to the events work (as I think that will require more changes to the endpoint).
* I think passing a server calculated session duration down makes sense. Let me know if you want help with that as I was just working on it.

**Comment by alexkim205** [comment]
Hey @rcmarron  retagging for a rereview. I addressed your feedback and added more tests (backend session recording + frontend sessionsPlayLogic) to capture some edge cases with this new way of requesting session recordings.

Changes from last review

- calculate duration on backend and pass through endpoint
- cache single session recordings with TTL 1min to help with long recordings that require multiple requests
- add error toast on failing to load a recording at any point
- add sessionsPlayLogic.test
- add and fix backend tests
- various mypy code smells that were causing Backend code quality checks to fail

**Review by rcmarron — CHANGES_REQUESTED** [review]
Just a couple of smaller comments! 

Also, I was struggling to get the updates from `posthog-react-rrweb-player` to work in the main `posthog` repo, so I couldn't fully test that out. I might need your help with that tomorrow.

**Comment by alexkim205** [comment]
Frontend checks won't pass until v1.1.3 of `https://github.com/PostHog/posthog-react-rrweb-player` is published on npm.

**Review by rcmarron — DISMISSED** [review]
Just spent some time trying to break this :-)

Found a couple of things
* First payload can still be large with 1000 snapshots (See the seconds request - 17mb). This happened on a recording that was ~1hr long, but nothing else outside the ordinary. The full snapshot (type 2) was around 700kb. And many of them were sent in the first 1000 snapshots.
image


* If a type:2 snapshot isn’t sent down in the first payload, the player never starts. You can repro this by looking in the payload where the first type-2 response is, and then setting the `RECORDINGS_NUM_SNAPSHOTS_LIMIT` to be less than that number.

* The player should adopt the timestamp from the data once it has a complete version of the recording. Right now, if the server sends down a duration that’s too short, the player reaches the end of the seek bar and keeps playing.

* If the player reaches the end of the buffered period, the player pauses, but then it doesn’t restart when more data arrives. Can repro this by setting the chunk size lower and moving the player to 16x speed.

* I think there’s something weird going on with request time (for me, it’s still ~3 seconds on subsequent requests even if the payload is small). With the caching, I’d expect it to be much faster (I verified the cache is being hit). Are you seeing payloads taking a while?

**Comment by alexkim205** [comment]
* That's really interesting, thanks for diving into the snapshot contents. In that case I wonder it's worth it to break up the snapshots in a "smarter" way. For example, we can always make sure that the first chunk returns contains at least one (~700kb) and at most ten (7MB) type 2 snapshots? A "dumb" but easy way to mitigate the issue would be to reduce the limit to 50 (or lower) snapshots and hope that the player will catch up eventually.
* First bullet would address that. 
* I think that makes sense. Displaying the total time as --:--:-- until recording is finished buffering will be a lot less confusing. I couldn't reproduce this specific instance of duration being shorter than snapshots indicate, but still better not to expose this inconsistency to the user. Should be fixed now.
* Fixed!
* Hm strange I'm running into this issue too. I timed a few session recordings and found that cache is working as expected. For example the first uncached request took 386ms to query while the cached request took close to 0s as expected. It seems like the bottleneck is actually in the decompressing of the chunks which consistently takes ~700ms per request. I avoided caching the decompressed data to avoid redis from overbloating, but I actually think the performance gains outweight the size limitations here, since we restrict cache TTL for each recording to 60s. I'll change the caching behavior to cache decompressed data for now and see if we see any major performance gains here.
    * Update: now caching decompressed data and it works much faster

---

@fuziontech Caching every session recording on viewing a recording can overload Redis clusters. We give it a TTL of only 60s here because each cache entry should have a lifetime of how long it takes to fetch an entire session recording in chunks. Are there other ways we can approach caching here that can reduce the burden on Redis but allow us to take advantage the cache?

**Comment by alexkim205** [comment]
Ok! Made some changes responding to feedback above @rcmarron :

- Now we're caching decompressed data instead of compressed chunks because the time bottleneck happened on decompress. 
- Change limit to 50 to make payloads smaller and likely for full snapshots to come quicker.
- Don't make player playable until we get a full snapshot. Added a test too
- Change cache TTL from 60s to 30s

**Comment by alexkim205** [comment]
I think it's worth getting this one in before it balloons into a bigger PR. To ensure this is a relatively safe merge, I've commented out caching with a TODO for now because we don't know what the impact will be on Redis and increased the limit per request from 50 to 100 to accommodate for this.

**Pull request merged**